lib/finch.ex

```diff
@@ -351,12 +350,84 @@ defmodule Finch do
 
   See `request/3` for more detailed information.
   """
-  @spec request!(Request.t(), name(), keyword()) ::
+  @spec request!(Request.t(), name(), request_opts()) ::
           Response.t()
   def request!(%Request{} = req, name, opts \\ []) do
     case request(req, name, opts) do
       {:ok, resp} -> resp
       {:error, exception} -> raise exception
     end
   end
+
+  @doc """
+  Sends an HTTP request asynchronously, returning a request reference.
+
+  ## Responses
+
+  Response information is sent to the calling process as it is received
+  in `{ref, response}` tuples. Responses include:
+
+    * `{:status, status}` - HTTP response status
+    * `{:headers, headers}` - HTTP response headers
+    * `{:data, data}` - section of the HTTP response body
+    * `{:error, exception}` - an error occurred during the request
+    * `:done` - request has completed successfully
+
+  On a successful request, a single `:status` message will be followed
+  by a single `:headers` message, after which one or more `:data`
+  messages may be sent. If trailing headers are present, it is possible
+  that a final `:headers` message may be sent. A `:done` or an `:error`
+  message indicates that the request has succeeded or failed and no
+  further messages are expected.
+
+  ## Example
+
+      iex> Stream.resource(
+      ...>   fn ->
+      ...>     Finch.build(:get, "https://httpbin.org/stream/5")
+      ...>     |> Finch.async_request(MyFinch)
+      ...>   end,
+      ...>   fn ref ->
+      ...>     receive do
+      ...>       {^ref, :done} -> {:halt, ref}
+      ...>       {^ref, response} -> {[response], ref}
+      ...>     end
+      ...>   end,
+      ...>   fn _ref -> :ok end
+      ...> ) |> Enum.to_list()
+      [
+        {:status, 200},
+        {:headers, [...]},
+        {:data, "..."},
+        ...
+        :done
+      ]
+
+  ## Options
+
+  Shares the same options as `request/3`.
+  """
+  @spec async_request(Request.t(), name(), request_opts()) :: request_ref()
+  def async_request(%Request{} = req, name, opts \\ []) do
+    {pool, pool_mod} = get_pool(req, name)
+    pool_mod.async_request(pool, req, opts)
+  end
+
+  @doc """
+  Cancels a request sent with `async_request/3`.
+  """
+  @spec cancel_async_request(request_ref()) :: :ok
+  def cancel_async_request(request_ref) when Finch.Pool.is_request_ref(request_ref) do
+    {_ref, _pool, pool_mod, _state} = request_ref
```

Comment:
I am worried that, if we have to keep any state, we now have to pass the copy of this state on every message. Maybe `async_request` should return `{stream_ref, cancel_ref}`?

Reply:
I don’t feel strongly about this. Right now, HTTP1 uses the state to store the pid of the spawned process wrapping the blocking `request` call, and HTTP2 doesn’t use the state at all (field is `nil`). 

So HTTP1 needs:
- pool mod to know it’s HTTP1
- PID of spawned process (will be unique per request, so we don’t even need a reference here)

HTTP2 needs:
- pool mod to know it’s HTTP2
- pool pid to know which pool process is handling the request/response
- reference to differentiate between requests

So if we wanted to simplify this value, it could be:

```elixir
module Finch.Pool do
  @type cancel_ref :: term()
 
  @callback async_request(…) :: cancel_ref()
  @callback cancel_async_request(cancel_ref()) :: :ok

  ...
end

defmodule Finch do
  @opaque request_ref :: {pool_mod :: module, Finch.Pool.cancel_ref()}
end

# HTTP1
{__MODULE__, pid}

# HTTP2
{__MODULE__, {pool_pid, make_ref()}}
```

And let each pool keep only what’s needed to cancel the request. This would, I think, better communicate to authors in the future that this shouldn’t be an arbitrary bunch of state, but only what’s needed to cancel an in-flight request.

Reply:
> So if we wanted to simplify this value, it could be:

Yes, this would be nice (and enough to keep me quiet). :)

Reply:
I never want you to stay quiet! 😄

Reply:
Done in f89106b926a3490f226466ee1d6c199236c90374, not quite exactly as above, but definitely simpler and allowed getting rid of `receive_next_within!` completely.